File: 0347 top-k-frequent-elements/solution.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        int n = nums.size();
        unordered_map<int, int> freq;
        for(int i:nums) freq[i]++;
        vector<vector<int>> bucket(n+1);
        for (auto &p : freq) bucket[p.second].push_back(p.first); 
        vector<int> ans;
        for(int i=n;i>=0;i--) {
            for(int num:bucket[i]) {
                ans.push_back(num);
                if (ans.size() == k) return ans;
            }
        }
        return ans;
    }
};

File: 0347 top-k-frequent-elements/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut fr: HashMap<i32, i32> = HashMap::new();
        for i in nums {
            *fr.entry(i).or_insert(0) += 1;
        }

        let max_freq = *fr.values().max().unwrap();

        let mut buc: Vec<Vec<i32>> = vec![Vec::new(); (max_freq + 1) as usize];
        for (num, f) in fr {
            buc[f as usize].push(num);
        }

        let mut ans: Vec<i32> = Vec::new();
        for freq in (0..buc.len()).rev() {
            for &num in &buc[freq] {
                if ans.len() < k as usize {
                    ans.push(num);
                } else {
                    break;
                }
            }
            if ans.len() == k as usize {
                break;
            }
        }

        ans
    }
}

